src/main/java/org/opentripplanner/routing/trippattern/TripTimes.java

```diff
@@ -359,6 +371,76 @@ public int getDropoffType(int stop) {
         return dropoffs[stop];
     }
 
+    /**
+     * Fetches pickup booking rule corresponding to the stop number in the trip
+     * @param stop - iterator in the trip
+     * @return pickup booking rule with all its properties
+     */
+    public BookingRule getPickupBookingRule(int stop) {
+        return pickupBookingRules[stop];
+    }
+
+    /**
+     * Fetches drop-off booking rule corresponding to the stop number in the trip
+     * @param stop - iterator in the trip
+     * @return drop-off booking rule with all its properties
+     */
+    public BookingRule getDropOffBookingRule(int stop) {
+        return dropOffBookingRules[stop];
+    }
+
+    /**
+     * Calculates the earliest pickup time based on a stop's pickup booking rule (GTFS-Flex v2.1)
+     * @param stop - iterator in the StopTimes collection
+     * @param currTime - current time
+     * @param sd - service day
+     * @return time
+     */
+    public long getEarliestPickupBookingTime (int stop, long currTime, ServiceDay sd) {
+        BookingRule pickupBookingRule = pickupBookingRules[stop];
+        int bookingRuleType = pickupBookingRule.getType();
+        switch (bookingRuleType) {
+            case 0:
+                // Real-time booking only
+                return currTime;
+            case 1:
+                // Up to same-day booking, with advance notice
+                return (long) (currTime + Math.round(pickupBookingRule.getPriorNoticeDurationMin() * 60.0));
+            case 2:
+                // Up to prior day(s) booking
+                int priorNoticeStartDay = pickupBookingRule.getPriorNoticeStartDay();
+                int priorNoticeStartTime = pickupBookingRule.getPriorNoticeStartTime();
+                return sd.time((int) Math.round((priorNoticeStartDay * 24.0 + priorNoticeStartTime) * 60.0));
+        }
+        return currTime;
+    }
+
+    /**
+     * Calculates the latest drop-off time based on a stop's pickup booking rule (GTFS-Flex v2.1)
+     * @param stop - iterator in the StopTimes collection
+     * @param currTime - current time
+     * @param sd - service day
+     * @return time
+     */
+    public long getLatestDropOffBookingTime (int stop, long currTime, ServiceDay sd) {
```

Comment:
Does this make sense. The booking rules, are always calculated from the boarding time, so the above method would bee needed to check that a trip was valid.